pytorch_lightning/loops/closure.py

```diff
@@ -11,32 +11,84 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
 from abc import ABC, abstractmethod
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from typing import Any, Callable, Dict, Optional
 
 from torch import Tensor
 
 from pytorch_lightning.profiler import BaseProfiler, PassThroughProfiler
-from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.utilities.apply_func import apply_to_collection
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.warnings import WarningCache
+from pytorch_lightning.utilities.memory import recursive_detach
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from pytorch_lightning.utilities.warnings import rank_zero_deprecation, WarningCache
 
 
 @dataclass
 class ClosureResult:
```

Comment:
When I introduced this, my intention was to keep the closure and closure result abstract concepts that extend to any loop structure and enables full loop customization. I intentionally removed as many references as possible to training_step or concrete implementations that would assume a certain loop structure or output format. That's why I have an abstract closure and a concrete closure definition below. 

With the changes here, this idea is now again threatened. While I very much like the refactors and simplification in the optimizer loop, I am not a fan of the direction the closure result is taking First, the API here suggests there is a training_step. Second, it assumes the output format to contain a loss and optionally hiddens or extras. This is working against loop customization where the user will need to return a ClosureResult but instead of implementing an interface they are overriding all  these methods just to prevent the default behavior.

I would like to suggest that we keep the interface separate from the implementation, meaning the closure result specific to the automatic optimization in lightning should be implemented together with our optimizer loop.

Reply:
> That's why I have an abstract closure and a concrete closure definition below.

I completely agree and note that the `AbstractClosure` and `Closure` are untouched (not counting the type annotations).

> First, the API here suggests there is a training_step

This was already the case for `ClosureResult` given it's fields, just that it was partly splitted there and in the `utilities` file.
In any case, the presence of a `training_step` is a given in Lightning.

> Second, it assumes the output format to contain a loss and optionally hiddens or extras

This logic needs to be defined somewhere, as they are the valid options for the optimizer/manual loop.

>  the closure result specific to the automatic optimization in lightning should be implemented together with our optimizer loop.

Agree, this is what we discussed in our call (see the summarization in the top post). However, this PR is already 200 lines and didn't want to expand its scope.

We can have the `OutputResult` as the abstract implementation of a training step result and a `ManualResult(OutputResult)` defined alongside the `ManualLoop`, as well as a `ClosureResult` defined alongside the `OptimizationLoop` which would be the one returned by the `Closure`

Reply:
If you prefer, I can try and add these shell classes first and then use them in this PR.

Reply:
> This was already the case for ClosureResult given it's fields

Because at that time manual and automatic opt was all still mangled together in one TrainingBatchLoop. It was very hard to get the logic for an "optional" result out there and what really helped here was splitting the loops for automatic and manual.  

> In any case, the presence of a training_step is a given in Lightning.

Not for loop customization. In a situation like #8643 the training step hook does not return the output directly but rather a generator. Then, in the closure step_fn is defined as a function that calls next on the generator. 


> If you prefer, I can try and add these shell classes first and then use them in this PR.

Or can you just give me some notes on what you will have the end result look like?

Reply:
> Or can you just give me some notes on what you will have the end result look like?

See #9424.

Reply:
resolving this as I'm ok with #9424 :)